### PR TITLE
feat: `Snapshotter` triggers segment snapshots

### PR DESCRIPTION
**Pull request by joshieDo**
PR into #5283 

* Adds trigger [`Snapshotter`] logic: `run_segment`
* Snapshot files are built in a temporary directory inside `datadir/snapshots/temp`, and moved to `datadir/snapshots` once finished. This is to handle abrupt terminations of the node. This temp directory is cleaned up whenever the node is started.

### CODE REVIEW

File: crates/snapshot/src/segments/mod.rs [hunk 1]
```diff
@@ -29,9 +31,13 @@ pub trait Segment {
     fn snapshot<DB: Database>(
```
Comment (mattsse) [inline crates/snapshot/src/segments/mod.rs:31]: this needs docs for directory.

why is this a pathbuf and not a &Path.

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/snapshot/src/snapshotter.rs [hunk 2]
```diff
@@ -17,6 +17,10 @@ pub type SnapshotterResult = Result<SnapshotTargets, SnapshotterError>;
 /// The snapshotter type itself with the result of [Snapshotter::run]
 pub type SnapshotterWithResult<DB> = (Snapshotter<DB>, SnapshotterResult);
 
+/// Snapshots are initially created in `{...}/datadir/snapshots/temp` and moved once finished. This
+/// directory is cleaned up on every booting up of the node.
+const TEMPORARY_SUBDIRECTORY: &str = "temp";
```
Comment (shekhirin) [inline crates/snapshot/src/snapshotter.rs:22]: why not system-wide `std::env::temp_dir`?

File: crates/snapshot/src/snapshotter.rs [hunk 3]
```diff
@@ -111,11 +105,35 @@ impl<DB: Database> Snapshotter<DB> {
...
+        if !self.snapshots_path.exists() {
+            std::fs::create_dir_all(&self.snapshots_path).map_err(err)?;
+        } else if temporary_path.exists() {
+            std::fs::remove_dir_all(&temporary_path).map_err(err)?;
+        }
+
+        std::fs::create_dir_all(temporary_path).map_err(err)?;
+
+        Ok(())
+    }
```
Comment (shekhirin) [inline crates/snapshot/src/snapshotter.rs:135]: we can make use of `reth_primitives::fs` here

File: crates/snapshot/src/snapshotter.rs [hunk 4]
```diff
@@ -173,13 +191,49 @@ impl<DB: Database> Snapshotter<DB> {
...
 
+    /// Run the snapshotter for one segment.
+    ///
+    /// It first builds the snapshot in a **temporary directory** inside the snapshots directory. If
+    /// for some reason the node is terminated during the snapshot process, it will be cleaned
+    /// up on boot (on [`Snapshotter::new`]) and the snapshot process restarted from scratch for
+    /// this block range and segment.
+    ///
+    /// If it succeeds, then we move the snapshot file from the temporary directory to its main one.
+    fn run_segment<S: Segment + Default>(
```
Comment (shekhirin) [inline crates/snapshot/src/snapshotter.rs:217]: should we just make `trait Segment: Default`?

**Review by shekhirin — COMMENTED** [review]
first pass with nits, reviewing the segment running logic now

File: crates/snapshot/src/snapshotter.rs (hunk 2 above)
Comment (joshieDo) [reply crates/snapshot/src/snapshotter.rs:22]: that would go into `/tmp/`, which the user might not be aware and might not have storage space for it and/or might be in a different partition.

With it being inside the datadir, we're at least sure that there is space

File: crates/snapshot/src/snapshotter.rs [hunk 5]
```diff
@@ -167,13 +189,49 @@ impl<DB: Database> Snapshotter<DB> {
...
+    /// for some reason the node is terminated during the snapshot process, it will be cleaned
+    /// up on boot (on [`Snapshotter::new`]) and the snapshot process restarted from scratch for
+    /// this block range and segment.
+    ///
+    /// If it succeeds, then we move the snapshot file from the temporary directory to its main one.
+    fn run_segment<S: Segment>(
+        &self,
+        block_range: Option<RangeInclusive<BlockNumber>>,
+    ) -> RethResult<()> {
+        if let Some(block_range) = block_range {
```
Comment (mattsse) [inline crates/snapshot/src/snapshotter.rs:219]: can use let Some else return here

**Review by mattsse — APPROVED** [review]
lgtm, pending @shekhirin

**Review by shekhirin — APPROVED** [review]

**Review by rkrasiuk — APPROVED** [review]
changes in isolation lgtm. still familiarizing with the rest of snapshot code

**Review by mattsse — APPROVED** [review]

**Pull request merged**